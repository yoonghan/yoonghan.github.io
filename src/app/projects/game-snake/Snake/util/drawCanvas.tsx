const drawSquareBoard = (
  ctx: CanvasRenderingContext2D,
  dimensionWidth: number,
  cellSize: number,
) => {
  ctx.beginPath()

  for (let x = 0; x <= dimensionWidth; x++) {
    ctx.moveTo(x * cellSize, 0)
    ctx.lineTo(x * cellSize, dimensionWidth * cellSize)
  }

  for (let y = 0; y <= dimensionWidth; y++) {
    ctx.moveTo(0, y * cellSize)
    ctx.lineTo(dimensionWidth * cellSize, y * cellSize)
  }

  ctx.stroke()
}

const drawCell = (
  ctx: CanvasRenderingContext2D,
  idx: number,
  dimensionWidth: number,
  cellSize: number,
  fillColor = "#ff0000",
) => {
  const col = idx % dimensionWidth
  const row = Math.floor(idx / dimensionWidth)

  ctx.fillStyle = fillColor

  ctx.beginPath()

  ctx.fillRect(col * cellSize, row * cellSize, cellSize, cellSize)

  ctx.stroke()
}

const clearBoard = (
  ctx: CanvasRenderingContext2D,
  canvas: HTMLCanvasElement,
) => {
  ctx.clearRect(0, 0, canvas.width, canvas.height)
}

export { drawSquareBoard, drawCell, clearBoard }
