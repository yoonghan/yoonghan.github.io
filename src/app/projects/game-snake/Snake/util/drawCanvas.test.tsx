import { render, screen } from "@testing-library/react"
import { clearBoard, drawCell, drawSquareBoard } from "./drawCanvas"
import * as canvasMock from "jest-canvas-mock"

describe("drawCanvas", () => {
  const getCanvasTestId = "canvas-test"

  const renderComponent = () => {
    render(<canvas data-testid={getCanvasTestId} />)
  }

  it("should draw a squared board", () => {
    renderComponent()
    const canvas = screen.getByTestId(getCanvasTestId) as HTMLCanvasElement
    const ctx = canvas.getContext("2d")!
    drawSquareBoard(ctx, 5, 5)
    const calls = ctx.__getDrawCalls()
    expect(calls).toMatchSnapshot()
  })

  it("should draw a cell", () => {
    renderComponent()
    const canvas = screen.getByTestId(getCanvasTestId) as HTMLCanvasElement
    const ctx = canvas.getContext("2d")!
    drawCell(ctx, 0, 10, 5)
    const calls = ctx.__getDrawCalls()
    expect(calls).toMatchSnapshot()
  })

  it("should be able to clear", () => {
    renderComponent()
    const canvas = screen.getByTestId(getCanvasTestId) as HTMLCanvasElement
    const ctx = canvas.getContext("2d")!
    drawCell(ctx, 0, 10, 5, "green")
    clearBoard(ctx, canvas)
    const calls = ctx.__getDrawCalls()
    expect(calls).toMatchSnapshot()
  })
})
