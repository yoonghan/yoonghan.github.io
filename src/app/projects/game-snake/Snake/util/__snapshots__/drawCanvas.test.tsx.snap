// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawCanvas should be able to clear 1`] = `
[
  {
    "props": {
      "height": 5,
      "width": 5,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  {
    "props": {
      "height": 150,
      "width": 300,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
]
`;

exports[`drawCanvas should draw a cell 1`] = `
[
  {
    "props": {
      "height": 5,
      "width": 5,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;

exports[`drawCanvas should draw a squared board 1`] = `
[
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        {
          "props": {
            "x": 0,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 0,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 5,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 5,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 10,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 10,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 15,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 15,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 20,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 20,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 25,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 25,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 0,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 25,
            "y": 0,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 0,
            "y": 5,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 25,
            "y": 5,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 0,
            "y": 10,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 25,
            "y": 10,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 0,
            "y": 15,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 25,
            "y": 15,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 0,
            "y": 20,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 25,
            "y": 20,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        {
          "props": {
            "x": 0,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 25,
            "y": 25,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;
